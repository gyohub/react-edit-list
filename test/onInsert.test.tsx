import React from 'react';
import {cleanup, fireEvent, render, waitFor} from '@testing-library/react';

import ReactEditList, * as REL from 'react-edit-list';

import {schema, data, onLoad} from './data';

describe('onInsert()', () => {
    it('onInsert() undefined', async () => {
        let recvItem: REL.Row = {};
        const onInsert = jest.fn((item: REL.Row): void => {
            recvItem = item;
        });
        const r = render(<ReactEditList schema={schema} onLoad={onLoad} onInsert={onInsert} />);
        await waitFor(() => expect(r.getByText(/Desk/)).toBeInTheDocument());
        expect(r.container.innerHTML).toMatchSnapshot();

        const cells = r.container.querySelectorAll('td');
        fireEvent(cells[cells.length - 5], new MouseEvent('click', {bubbles: true}));
        await waitFor(() => expect(r.container.querySelectorAll('input').length).toBe(3));
        expect(r.container.innerHTML).toMatchSnapshot();

        const inputs = r.container.querySelectorAll('input');
        fireEvent.change(inputs[0], {target: {value: 'test item'}});
        fireEvent(await r.findByText('✓'), new MouseEvent('click', {bubbles: true}));
        expect(onInsert).toBeCalledTimes(1);

        await waitFor(() => expect(recvItem.product).toBe('test item'));
        expect(recvItem.type).toBeUndefined();
        expect(recvItem.price).toBeUndefined();
        expect(recvItem.stock).toBeUndefined();
        await waitFor(() => expect(r.container.querySelectorAll('input')).toHaveLength(0));
        expect(r.container.innerHTML).toMatchSnapshot();
        r.unmount();
    });

    it('onInsert() false', async () => {
        let recvItem: REL.Row = {};
        const onInsert = jest.fn((item: REL.Row): Promise<false> => {
            recvItem = item;
            return Promise.resolve(false);
        });
        const r = render(<ReactEditList schema={schema} onLoad={onLoad} onInsert={onInsert} />);
        await waitFor(() => expect(r.getByText(/Desk/)));
        expect(r.container.innerHTML).toMatchSnapshot();

        const cells = r.container.querySelectorAll('td');
        fireEvent(cells[cells.length - 5], new MouseEvent('click', {bubbles: true}));
        await waitFor(() => expect(r.container.querySelectorAll('input').length).toBe(3));
        expect(r.container.innerHTML).toMatchSnapshot();

        const inputs = r.container.querySelectorAll('input');
        fireEvent.change(inputs[0], {target: {value: 'test item 2'}});
        fireEvent(await r.findByText('✓'), new MouseEvent('click', {bubbles: true}));
        expect(onInsert).toBeCalledTimes(1);

        await waitFor(() => expect(recvItem.product).toBe('test item 2'));
        expect(recvItem.type).toBeUndefined();
        expect(recvItem.price).toBeUndefined();
        expect(recvItem.stock).toBeUndefined();
        await waitFor(() => expect(r.container.querySelectorAll('input')).toHaveLength(3));
        expect(r.container.innerHTML).toMatchSnapshot();
        r.unmount();
    });

    it('onInsert() false', async () => {
        let recvItem: REL.Row = {};
        const onInsert = jest.fn((item: REL.Row): Promise<REL.Row> => {
            recvItem = item;
            return Promise.resolve({product: 'something different'});
        });
        const r = render(<ReactEditList schema={schema} onLoad={onLoad} onInsert={onInsert} />);
        await waitFor(() => expect(r.getByText(/Desk/)));
        expect(r.container.innerHTML).toMatchSnapshot();

        const cells = r.container.querySelectorAll('td');
        fireEvent(cells[cells.length - 5], new MouseEvent('click', {bubbles: true}));
        await waitFor(() => expect(r.container.querySelectorAll('input').length).toBe(3));
        expect(r.container.innerHTML).toMatchSnapshot();

        const inputs = r.container.querySelectorAll('input');
        fireEvent.change(inputs[0], {target: {value: 'test item 3'}});
        fireEvent(await r.findByText('✓'), new MouseEvent('click', {bubbles: true}));
        expect(onInsert).toBeCalledTimes(1);

        await waitFor(() => expect(recvItem.product).toBe('test item 3'));
        expect(recvItem.type).toBeUndefined();
        expect(recvItem.price).toBeUndefined();
        expect(recvItem.stock).toBeUndefined();
        await waitFor(() => expect(r.container.querySelectorAll('input')).toHaveLength(0));
        await waitFor(() => expect(r.getByText(/something different/)));
        expect(r.container.innerHTML).toMatchSnapshot();
        r.unmount();
    });

    it('disableInsert', async () => {
        const onLoadFn = jest.fn(onLoad);
        const r = render(<ReactEditList schema={schema} onLoad={onLoadFn} disableInsert={true} />);
        await waitFor(() => expect(r.getByText(/Desk/)));
        expect(onLoadFn).toBeCalledTimes(1);
        expect(r.container.innerHTML).toMatchSnapshot();
        const cells = r.container.querySelectorAll('td');
        fireEvent(cells[cells.length - 5], new MouseEvent('click', {bubbles: true}));
        await waitFor(() => expect(r.container.querySelectorAll('input').length).toBe(3));
        expect(cells[cells.length - 5].children[0]).toHaveAttribute('value', 'Rent');
        r.unmount();
    });

    it('onInsert() defaultValues and filler', async () => {
        let recvItem: REL.Row = {};
        const onInsert = jest.fn((item: REL.Row): void => {
            recvItem = item;
        });
        const r = render(
            <ReactEditList
                schema={schema}
                onLoad={onLoad}
                onInsert={onInsert}
                defaultValues={{stock: 42}}
                filler={<span>...</span>}
            />
        );
        await waitFor(() => expect(r.getByText(/Desk/)).toBeInTheDocument());
        expect(r.container.innerHTML).toMatchSnapshot();

        fireEvent((await r.findAllByText('...'))[0], new MouseEvent('click', {bubbles: true}));
        await waitFor(() => expect(r.container.querySelectorAll('input').length).toBe(3));
        expect(r.container.innerHTML).toMatchSnapshot();

        const inputs = r.container.querySelectorAll('input');
        fireEvent.change(inputs[0], {target: {value: 'default item'}});
        fireEvent(await r.findByText('✓'), new MouseEvent('click', {bubbles: true}));
        expect(onInsert).toBeCalledTimes(1);

        await waitFor(() => expect(recvItem.product).toBe('default item'));
        expect(recvItem.type).toBeUndefined();
        expect(recvItem.price).toBeUndefined();
        expect(recvItem.stock).toBe(42);
        await waitFor(() => expect(r.container.querySelectorAll('input')).toHaveLength(0));
        expect(r.container.innerHTML).toMatchSnapshot();
        r.unmount();
    });
});
