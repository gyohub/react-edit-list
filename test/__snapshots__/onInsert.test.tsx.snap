// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`onInsert() onInsert() false 1`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>Desk</td><td>capex</td><td>100</td><td>20</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td></td><td></td><td></td><td></td><td><div style=\\"display: flex; flex-direction: row;\\">&nbsp;</div></td></tr></tbody></table>"`;

exports[`onInsert() onInsert() false 2`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>Desk</td><td>capex</td><td>100</td><td>20</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td><input type=\\"text\\" value=\\"\\"></td><td><select><option></option><option value=\\"1\\">capex</option><option value=\\"2\\">consumable</option></select></td><td><input type=\\"number\\" value=\\"\\"></td><td><input type=\\"number\\" value=\\"\\"></td><td><div style=\\"display: flex; flex-direction: row;\\"><button>✓</button><button>x</button></div></td></tr></tbody></table>"`;

exports[`onInsert() onInsert() false 3`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>Desk</td><td>capex</td><td>100</td><td>20</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td><input type=\\"text\\" value=\\"test item 2\\"></td><td><select><option></option><option value=\\"1\\">capex</option><option value=\\"2\\">consumable</option></select></td><td><input type=\\"number\\" value=\\"\\"></td><td><input type=\\"number\\" value=\\"\\"></td><td><div style=\\"display: flex; flex-direction: row;\\"><button>✓</button><button>x</button></div></td></tr></tbody></table>"`;

exports[`onInsert() onInsert() false 4`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>Desk</td><td>capex</td><td>100</td><td>20</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td></td><td></td><td></td><td></td><td><div style=\\"display: flex; flex-direction: row;\\">&nbsp;</div></td></tr></tbody></table>"`;

exports[`onInsert() onInsert() false 5`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>Desk</td><td>capex</td><td>100</td><td>20</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td><input type=\\"text\\" value=\\"\\"></td><td><select><option></option><option value=\\"1\\">capex</option><option value=\\"2\\">consumable</option></select></td><td><input type=\\"number\\" value=\\"\\"></td><td><input type=\\"number\\" value=\\"\\"></td><td><div style=\\"display: flex; flex-direction: row;\\"><button>✓</button><button>x</button></div></td></tr></tbody></table>"`;

exports[`onInsert() onInsert() false 6`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>Desk</td><td>capex</td><td>100</td><td>20</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>something different</td><td></td><td></td><td></td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td></td><td></td><td></td><td></td><td><div style=\\"display: flex; flex-direction: row;\\">&nbsp;</div></td></tr></tbody></table>"`;

exports[`onInsert() onInsert() undefined 1`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>Desk</td><td>capex</td><td>100</td><td>20</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td></td><td></td><td></td><td></td><td><div style=\\"display: flex; flex-direction: row;\\">&nbsp;</div></td></tr></tbody></table>"`;

exports[`onInsert() onInsert() undefined 2`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>Desk</td><td>capex</td><td>100</td><td>20</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td><input type=\\"text\\" value=\\"\\"></td><td><select><option></option><option value=\\"1\\">capex</option><option value=\\"2\\">consumable</option></select></td><td><input type=\\"number\\" value=\\"\\"></td><td><input type=\\"number\\" value=\\"\\"></td><td><div style=\\"display: flex; flex-direction: row;\\"><button>✓</button><button>x</button></div></td></tr></tbody></table>"`;

exports[`onInsert() onInsert() undefined 3`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>Desk</td><td>capex</td><td>100</td><td>20</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>test item</td><td></td><td></td><td></td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td></td><td></td><td></td><td></td><td><div style=\\"display: flex; flex-direction: row;\\">&nbsp;</div></td></tr></tbody></table>"`;
